### ISSUE TITLE
[dev.icinga.com #10265] Duplicate key error on uq_customvariables in IDO pgsql

### ISSUE THREAD

**Issue by icinga-migration** [Icinga/icinga2#3465]
This issue has been migrated from Redmine: https://dev.icinga.com/issues/10265

**Created by adamaflynn on 2015-10-02 03:19:17 +00:00**

Assignee: _(none)_
Status: _Closed_  (closed on _2015-10-21 08:30:02 +00:00_)
Target Version: _(none)_
Last Update: _2015-10-21 08:30:02 +00:00 (in Redmine)_

    Icinga Version: 2.3.8
    Backport?: Not yet backported
    Include in Changelog: 1

---

Hi,

When I update my Icinga2 configuration and restart the daemon, sometimes I get duplicate key errors, followed by deadlocks. This manifests in the web UI as an error saying the backend isn't available (under System). It can also take a long time for config file changes to show up in the web UI after being made in the config files (several minutes).

I manage the config files on my Icinga2 cluster using Chef and both nodes have configs that are updated at mostly the same time (although I've confirmed the issues happens when I restart serially). The cluster has &gt;10k services, so a race between the two nodes, especially around updates, seems plausible. The error tends to only happen on one of the cluster members at a time.

Error in icinga.log:

    [2015-10-01 20:05:36 -0700] critical/IdoPgsqlConnection: Error "ERROR:  duplicate key value violates unique constraint "uq_customvariables"
    DETAIL:  Key (object_id, config_type, varname)=(9244, 1, bgp_peer) already exists.
    " when executing query "INSERT INTO icinga_customvariables (config_type, has_been_modified, instance_id, is_json, object_id, varname, varvalue) VALUES (E'1', E'0', 1, E'0', 9244, E'bgp_peer', E'$host.vars.internal_bgp_peer$')"
    Context:
            (0) Reconnecting to PostgreSQL IDO database 'ido-pgsql'

    [2015-10-01 20:05:36 -0700] warning/IdoPgsqlConnection: Exception during database operation: Verify that your database is operational!
    [2015-10-01 20:05:36 -0700] information/IdoPgsqlConnection: pgSQL IDO instance id: 1 (schema version: '1.13.0')
    [2015-10-01 20:05:44 -0700] critical/IdoPgsqlConnection: Error "ERROR:  deadlock detected
    DETAIL:  Process 14028 waits for ShareLock on transaction 5650847; blocked by process 12714.
    Process 12714 waits for ShareLock on transaction 5650832; blocked by process 14028.
    HINT:  See server log for query details.
    CONTEXT:  while updating tuple (46,25) in relation "icinga_hoststatus"
    " when executing query "UPDATE icinga_hoststatus SET acknowledgement_type = E'0',  active_checks_enabled = E'1',  check_command = E'hostalive',  check_source = E'icinga-a',  check_type = E'0',  current_check_attempt = E'1',  current_notification_number = E'0',  current_state = E'0',  endpoint_object_id = 10053,  event_handler = E'',  event_handler_enabled = E'1',  execution_time = E'4.0038051605224609',  flap_detection_enabled = E'1',  has_been_checked = E'1',  host_object_id = 1201,  instance_id = 1,  is_flapping = E'0',  is_reachable = E'1',  last_check = TO_TIMESTAMP(1443754922),  last_hard_state = E'0',  last_hard_state_change = TO_TIMESTAMP(1439252779),  last_state_change = TO_TIMESTAMP(1439252779),  last_time_up = TO_TIMESTAMP(1443754971),  latency = E'0',  long_output = E'',  max_check_attempts = E'4',  modified_host_attributes = E'0',  next_check = TO_TIMESTAMP(1443755164),  next_notification = TO_TIMESTAMP(1443755212),  normal_check_interval = E'1',  notifications_enabled = E'1',  output = E'PING OK - Packet loss = 0%, RTA = 0.59 ms',  passive_checks_enabled = E'1',  percent_state_change = E'0',  perfdata = E'rta=0.587000ms;3000.000000;5000.000000;0.000000 pl=0%;80;100;0',  problem_has_been_acknowledged = E'0',  process_performance_data = E'1',  retry_check_interval = E'0.25',  scheduled_downtime_depth = E'0',  should_be_scheduled = E'1',  state_type = E'1',  status_update_time = TO_TIMESTAMP(1443755143) WHERE host_object_id = 1201"
    Context:
            (0) Reconnecting to PostgreSQL IDO database 'ido-pgsql'
    [2015-10-01 20:05:44 -0700] warning/IdoPgsqlConnection: Exception during database operation: Verify that your database is operational!
    [2015-10-01 20:05:45 -0700] information/IdoPgsqlConnection: pgSQL IDO instance id: 1 (schema version: '1.13.0')
    [2015-10-01 20:05:52 -0700] critical/IdoPgsqlConnection: Error "ERROR:  deadlock detected
    DETAIL:  Process 14084 waits for ShareLock on transaction 5650868; blocked by process 12714.
    Process 12714 waits for ShareLock on transaction 5650851; blocked by process 14084.
    HINT:  See server log for query details.
    CONTEXT:  while updating tuple (64,3) in relation "icinga_hoststatus"
    " when executing query "UPDATE icinga_hoststatus SET acknowledgement_type = E'0',  active_checks_enabled = E'1',  check_command = E'hostalive',  check_source = E'icinga-a',  check_type = E'0',  current_check_attempt = E'1',  current_notification_number = E'0',  current_state = E'0',  endpoint_object_id = 10053,  event_handler = E'',  event_handler_enabled = E'1',  execution_time = E'3.9999959468841553',  flap_detection_enabled = E'1',  has_been_checked = E'1',  host_object_id = 1200,  instance_id = 1,  is_flapping = E'0',  is_reachable = E'1',  last_check = TO_TIMESTAMP(1443754935),  last_hard_state = E'0',  last_hard_state_change = TO_TIMESTAMP(1439252779),  last_state_change = TO_TIMESTAMP(1439252779),  last_time_up = TO_TIMESTAMP(1443754975),  latency = E'0',  long_output = E'',  max_check_attempts = E'4',  modified_host_attributes = E'0',  next_check = TO_TIMESTAMP(1443755199),  next_notification = TO_TIMESTAMP(1443755212),  normal_check_interval = E'1',  notifications_enabled = E'1',  output = E'PING OK - Packet loss = 0%, RTA = 0.30 ms',  passive_checks_enabled = E'1',  percent_state_change = E'0',  perfdata = E'rta=0.298000ms;3000.000000;5000.000000;0.000000 pl=0%;80;100;0',  problem_has_been_acknowledged = E'0',  process_performance_data = E'1',  retry_check_interval = E'0.25',  scheduled_downtime_depth = E'0',  should_be_scheduled = E'1',  state_type = E'1',  status_update_time = TO_TIMESTAMP(1443755151) WHERE host_object_id = 1200"
    Context:
            (0) Reconnecting to PostgreSQL IDO database 'ido-pgsql'
    [2015-10-01 20:05:52 -0700] warning/IdoPgsqlConnection: Exception during database operation: Verify that your database is operational!

icinga2 -V:

    icinga2 - The Icinga 2 network monitoring daemon (version: v2.3.8)

    Copyright (c) 2012-2015 Icinga Development Team (https://www.icinga.org)
    License GPLv2+: GNU GPL version 2 or later 
    This is free software: you are free to change and redistribute it.
    There is NO WARRANTY, to the extent permitted by law.

    Application information:
      Installation root: /usr
      Sysconf directory: /etc
      Run directory: /run
      Local state directory: /var
      Package data directory: /usr/share/icinga2
      State path: /var/lib/icinga2/icinga2.state
      Objects path: /var/cache/icinga2/icinga2.debug
      Vars path: /var/cache/icinga2/icinga2.vars
      PID path: /run/icinga2/icinga2.pid
      Application type: icinga/IcingaApplication

    System information:
      Operating system: Linux
      Operating system version: 3.2.0-75-virtual
      Architecture: x86_64
      Distribution: Ubuntu 12.04.5 LTS

zones.conf:

    object Endpoint "icinga-a" {
      host = "icinga-a.i.wish.com"
    }

    object Endpoint "icinga-c" {
      host = "icinga-c.i.wish.com"
    }

    object Zone "icinga-ha" {
      endpoints = [ "icinga-a", "icinga-c" ]
    }

    object Zone "global" {
      global = true
    }

---

**Relations**:

* relates [#10190](https://dev.icinga.com/issues/10190)
* relates [#10232](https://dev.icinga.com/issues/10232)
